clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerRebalanceListenerInvoker.java

```diff
@@ -0,0 +1,128 @@
+ * interface. When consumer group partition assignment changes, these methods are invoked. This class wraps those
+ * callback calls with some logging, optional {@link Sensor} updates, etc.
+ */
+class ConsumerRebalanceListenerInvoker {
+
+    private final Logger log;
+    private final SubscriptionState subscriptions;
+    private final Time time;
+    private final ConsumerCoordinatorMetrics coordinatorMetrics;
+
+    ConsumerRebalanceListenerInvoker(LogContext logContext,
+                                     SubscriptionState subscriptions,
+                                     Time time,
+                                     ConsumerCoordinatorMetrics coordinatorMetrics) {
+        this.log = logContext.logger(getClass());
+        this.subscriptions = subscriptions;
+        this.time = time;
+        this.coordinatorMetrics = coordinatorMetrics;
+    }
+
+    Exception invokePartitionsAssigned(final SortedSet<TopicPartition> assignedPartitions) {
+        log.info("Adding newly assigned partitions: {}", Utils.join(assignedPartitions, ", "));
+
+        Optional<ConsumerRebalanceListener> listener = subscriptions.rebalanceListener();
+
+        if (listener.isPresent()) {
+            try {
+                final long startMs = time.milliseconds();
+                listener.get().onPartitionsAssigned(assignedPartitions);
+                coordinatorMetrics.assignCallbackSensor.record(time.milliseconds() - startMs);
+            } catch (WakeupException | InterruptException e) {
+                throw e;
+            } catch (Exception e) {
+                log.error("User provided listener {} failed on invocation of onPartitionsAssigned for partitions {}",
+                        listener.getClass().getName(), assignedPartitions, e);
+                return e;
+            }
+        }
+
+        return null;
+    }
+
+    Exception invokePartitionsRevoked(final SortedSet<TopicPartition> revokedPartitions) {
+        log.info("Revoke previously assigned partitions {}", Utils.join(revokedPartitions, ", "));
+        Set<TopicPartition> revokePausedPartitions = subscriptions.pausedPartitions();
+        revokePausedPartitions.retainAll(revokedPartitions);
+        if (!revokePausedPartitions.isEmpty())
+            log.info("The pause flag in partitions [{}] will be removed due to revocation.", Utils.join(revokePausedPartitions, ", "));
+
+        Optional<ConsumerRebalanceListener> listener = subscriptions.rebalanceListener();
+
+        if (listener.isPresent()) {
+            try {
+                final long startMs = time.milliseconds();
+                listener.get().onPartitionsRevoked(revokedPartitions);
+                coordinatorMetrics.revokeCallbackSensor.record(time.milliseconds() - startMs);
+            } catch (WakeupException | InterruptException e) {
+                throw e;
+            } catch (Exception e) {
```

Comment:
we should definitely return all KafkaExceptions; are there java exceptions that we want to throw back immediately to the user though?

Reply:
I can't speak to all the different use cases, but this is what is done in the existing `ConsumerCoordinator`, so this was done to keep the existing behavior equivalent.